velox/dwio/parquet/reader/CompressionQpl.cpp

```diff
@@ -0,0 +1,207 @@
+    expected = false;
+    index = Getindex();
+  }
+
+  job_ = job_pool[index];
+  return index;
+}
+
+/*
+ * The QPL can't compress/decompress the block over QPL_MAX_TRANS_SIZE,
+ * We partition the input to several 256KB blocks.
+ * So that QPL can compress/decompress with a high throughput.
+ */
+bool Qplcodec::Compress(
+    int64_t input_length,
+    const uint8_t* input,
+    int64_t output_buffer_length,
+    uint8_t* output) {
+  int job_id = Getjob();
+  int64_t out_size = output_buffer_length;
+  job_->total_out = 0;
+  job_->total_in = 0;
+
+  job_->op = qpl_op_compress;
+  job_->level = (qpl_compression_levels)compression_level_;
+  job_->next_out_ptr = output;
+  job_->next_in_ptr = const_cast<uint8_t*>(input);
+  job_->flags =
+      QPL_FLAG_FIRST | QPL_FLAG_DYNAMIC_HUFFMAN | QPL_FLAG_OMIT_VERIFY;
+
+  int64_t current_chunk_size = MAX_COMPRESS_CHUNK_SIZE;
+  uint32_t iteration_count = 0;
+  int64_t source_bytes_left = input_length;
+
+  while (source_bytes_left > 0) {
+    if (current_chunk_size >= source_bytes_left) {
+      job_->flags |= QPL_FLAG_LAST;
+      current_chunk_size = source_bytes_left;
+    }
+
+    job_->available_in = current_chunk_size;
+    job_->available_out = QPL_MAX_TRANS_SIZE < out_size - job_->total_out
+        ? QPL_MAX_TRANS_SIZE
+        : out_size - job_->total_out;
+
+    qpl_status status = qpl_execute_job(job_);
+
+    if (status != QPL_STS_OK) {
+      std::atomic_store(&job_status[job_id], false);
+      VELOX_FAIL("QPL::Error while QPL compression occurred.");
+      return false;
+    }
+
+    source_bytes_left = input_length - job_->total_in;
+    job_->flags &= ~QPL_FLAG_FIRST;
+    iteration_count++;
+  }
+  std::atomic_store(&job_status[job_id], false);
+  return true;
+}
+
+bool Qplcodec::Decompress(
+    int64_t input_length,
+    const uint8_t* input,
+    int64_t output_buffer_length,
+    uint8_t* output) {
+  if (output_buffer_length == 0) {
+    return false;
+  }
+
+  int64_t out_size = output_buffer_length;
+  int job_id = Getjob();
+  job_->total_out = 0;
+  job_->total_in = 0;
+  job_->op = qpl_op_decompress;
+  job_->next_out_ptr = output;
+  job_->next_in_ptr = const_cast<uint8_t*>(input);
+  job_->flags = QPL_FLAG_FIRST;
+
+  int64_t current_chunk_size = MAX_DECOMPRESS_CHUNK_SIZE;
+  uint32_t iteration_count = 0;
+  int64_t source_bytes_left = input_length;
+
+  while (source_bytes_left > 0) {
+    if (current_chunk_size >= source_bytes_left) {
+      job_->flags |= QPL_FLAG_LAST;
+      current_chunk_size = source_bytes_left;
+    }
+    source_bytes_left -= current_chunk_size;
+    job_->available_in = current_chunk_size;
+    job_->available_out = QPL_MAX_TRANS_SIZE < out_size - job_->total_out
+        ? QPL_MAX_TRANS_SIZE
+        : out_size - job_->total_out;
+
+    qpl_status status = qpl_execute_job(job_);
+    if (status != QPL_STS_OK) {
+      std::atomic_store(&job_status[job_id], false);
+      VELOX_FAIL("QPL::Error while decompression occurred.");
```

Comment:
We need to choose whether to return the success state or using exception, not both.  Can you
1. Choose whether we return value (remove all throws in this file and throw up in the stack), or just throw here (return `void` instead)
2. Decide if empty buffer is a valid case?  If not, we need to throw an exception for that (here or up in the stack)

Reply:
Yeah, I have changed bool-return function to void function. And we throw an exception if buffer is empty.